cpp/src/strings/json/json_path.cu

```diff
@@ -0,0 +1,983 @@
+                                       char* out_buf,
+                                       bitmask_type* out_validity)
+{
+  uint64_t const tid = threadIdx.x + (blockDim.x * blockIdx.x);
+
+  bool is_valid = false;
+  if (tid < col.size()) {
+    string_view const str = col.element<string_view>(tid);
+    size_type output_size = 0;
+    if (str.size_bytes() > 0) {
+      char* dst             = out_buf ? out_buf + output_offsets[tid] : nullptr;
+      size_t const dst_size = out_buf ? output_offsets[tid + 1] - output_offsets[tid] : 0;
+
+      parse_result result;
+      json_output out;
+      thrust::tie(result, out) =
+        get_json_object_single(str.data(), str.size_bytes(), commands, dst, dst_size);
+      output_size = out.output_len;
+      if (out.has_output > 0 && result == parse_result::SUCCESS) { is_valid = true; }
+    }
+
+    // filled in only during the precompute step
+    if (!out_buf) { output_offsets[tid] = static_cast<size_type>(output_size); }
+  }
+
+  // validity filled in only during the output step
+  if (out_validity) {
+    uint32_t mask = __ballot_sync(0xffffffff, is_valid);
+    // 0th lane of the warp writes the validity
+    if (!(tid % cudf::detail::warp_size) && tid < col.size()) {
+      out_validity[cudf::word_index(tid)] = mask;
+    }
+  }
+}
+
+/**
+ * @copydoc cudf::strings::detail::get_json_object
+ */
+std::unique_ptr<cudf::column> get_json_object(cudf::strings_column_view const& col,
+                                              cudf::string_scalar const& json_path,
+                                              rmm::cuda_stream_view stream,
+                                              rmm::mr::device_memory_resource* mr)
+{
+  // preprocess the json_path into a command buffer
+  std::tuple<rmm::device_uvector<path_operator>, int, bool> preprocess =
+    build_command_buffer(json_path, stream);
+  CUDF_EXPECTS(std::get<1>(preprocess) <= max_command_stack_depth,
+               "Encountered json_path string that is too complex");
+
+  // allocate output offsets buffer.
+  auto offsets = cudf::make_fixed_width_column(
+    data_type{type_id::INT32}, col.size() + 1, mask_state::UNALLOCATED, stream, mr);
+  cudf::mutable_column_view offsets_view(*offsets);
+
+  // if the query is empty, return a string column containing all nulls
+  if (std::get<2>(preprocess)) {
+    return std::make_unique<column>(
+      data_type{type_id::STRING},
+      col.size(),
+      rmm::device_buffer{0, stream, mr},  // no data
+      cudf::detail::create_null_mask(col.size(), mask_state::ALL_NULL, stream, mr),
+      col.size());  // null count
+  }
+
+  cudf::detail::grid_1d const grid{col.size(), 512};
+
+  auto cdv = column_device_view::create(col.parent(), stream);
+
+  // preprocess sizes (returned in the offsets buffer)
+  get_json_object_kernel<<<grid.num_blocks, grid.num_threads_per_block, 0, stream.value()>>>(
+    *cdv, std::get<0>(preprocess).data(), offsets_view.head<size_type>(), nullptr, nullptr);
```

Comment:
One reason for using a simple `thrust::for_each` instead of launching a basic C-function kernel is that `thrust` will try to set the grid/block size appropriately for the target architecture so we don't have to hardcode the block size here. Also your kernel does not need to include a boundary check `if( tid < col.size())`.